sql/uniques.h

```diff
@@ -38,45 +246,118 @@ class Unique :public Sql_alloc
+                                always 0 for unions, > 0 for intersections */
+  const bool with_counters;
+
+  // size in bytes used for storing keys in the Unique tree
+  size_t memory_used;
+  ulong elements;
+  SORT_INFO sort;
+
+  /*
+    Storing all meta-data information of the expressions whose value are
+    being added to the Unique tree
+  */
+  Keys_descriptor *keys_descriptor;
 
   bool merge(TABLE *table, uchar *buff, size_t size, bool without_last_merge);
   bool flush();
 
-public:
-  ulong elements;
-  SORT_INFO sort;
-  Unique(qsort_cmp2 comp_func, void *comp_func_fixed_arg,
-	 uint size_arg, size_t max_in_memory_size_arg,
-         uint min_dupl_count_arg= 0);
-  ~Unique();
-  ulong elements_in_tree() { return tree.elements_in_tree; }
-  inline bool unique_add(void *ptr)
+  // return the amount of unused memory in the Unique tree
+  size_t space_left() const
+  {
+    DBUG_ASSERT(max_in_memory_size >= memory_used);
+    return max_in_memory_size - memory_used;
+  }
+
+  // Check if the Unique tree is full or not
+  bool is_full(size_t record_size)
+  {
+    if (!tree.elements_in_tree)  // Atleast insert one element in the tree
+      return false;
+    return record_size > space_left();
+  }
+
+  /*
+    @brief
+      Add a record to the Unique tree
+    @param
+      ptr                      key value
+      size                     length of the key
+    @retval
+      TRUE                     ERROE
+      FALSE                    key successfully inserted in the Unique tree
+  */
+
+  bool unique_add(void *ptr, uint key_size)
   {
     DBUG_ENTER("unique_add");
     DBUG_PRINT("info", ("tree %u - %lu", tree.elements_in_tree, max_elements));
-    if (!(tree.flag & TREE_ONLY_DUPS) && 
-        tree.elements_in_tree >= max_elements && flush())
+    TREE_ELEMENT *res;
+    size_t rec_size= key_size + sizeof(TREE_ELEMENT) + tree.size_of_element;
+
+    if (!(tree.flag & TREE_ONLY_DUPS) && is_full(rec_size) && flush())
       DBUG_RETURN(1);
-    DBUG_RETURN(!tree_insert(&tree, ptr, 0, tree.custom_arg));
+    uint count= tree.elements_in_tree;
+    res= tree_insert(&tree, ptr, key_size, tree.custom_arg);
+    if (tree.elements_in_tree != count)
+    {
+      /*
+        increment memory used only when a unique element is inserted
+        in the tree
+      */
+      memory_used+= rec_size;
+    }
+    DBUG_RETURN(!res);
+  }
+
+public:
+
+  /*
+    @brief
+      Returns the number of elements in the unique instance
+
+    @details
+      If all the elements fit in the memory, then this returns all the
+      distinct elements.
+  */
+  ulong get_n_elements()
+  {
+    return is_in_memory() ? elements_in_tree() : elements;
+  }
+
+  SORT_INFO *get_sort() { return &sort; }
+
+  Unique(Keys_descriptor *desc,
+         size_t max_in_memory_size_arg,
+         uint min_dupl_count_arg);
+  ~Unique();
+  ulong elements_in_tree() { return tree.elements_in_tree; }
+
+  bool unique_add(uchar *ptr)
```

Comment:
So, now there are two methods:
```
private:
    bool unique_add(void *ptr, uint key_size);
public:
    bool unique_add(uchar *ptr);
```

Am I the only one who thinks it's confusing?

Reply:
Actually, they have different meanings.
`bool unique_add(uchar *ptr)` - this one accepts the in some "source format".   It uses `keys_descriptor` to convert it to some internal representation and then `bool unique_add(void *ptr, uint key_size);`  is called to insert that into the RB-tree...

Reply:
As a user of the unique class, you don't have to care conceptually about the unique_add method with two parameters.

It's part of the internal implementation. If you think this would be better named unique_add_internal or add_internal or something else, I could change it to that, but I consider the fact that it's marked private a sufficiently strong of a hint.

When writing this code I took two stances: as a class implementer, the names of private methods are of no concern to the user of said class, at all. I didn't care about any conflicts.
As a user of a class, the code strives to rely strictly on the public methods, not rely on what they internally do. There were cases when Varun's patch required a certain order of methods being called for them to work. I tried to minimize that.

Reply:
Also, one more thing: the unique object (albeit not completely clear cut with this current iteration) doesn't actually handle the "internal representation format". That part is delegated to the Keys_descriptor. The unique object just handles storing stuff in the tree, optionally on disk when memory limit exceeded and getting it back out. The other details of fields, packed values, etc. should be handled by the Keys_descriptor.